source/common/grpc/google_grpc_creds_impl.cc

```diff
@@ -140,7 +140,7 @@ class DefaultGoogleGrpcCredentialsFactory : public GoogleGrpcCredentialsFactory
 public:
   std::shared_ptr<grpc::ChannelCredentials>
   getChannelCredentials(const envoy::config::core::v3::GrpcService& grpc_service_config,
-                        Api::Api& api) override {
+                        Api::Api& api, Upstream::ClusterManager&) override {
```

Comment:
Yes all these effort is to get the cluster manager inside source/extensions/grpc_credentials/aws_iam/config.cc

It will eventually be used in my follow up CRs to create internal cluster 

For example like here: https://github.com/suniltheta/envoy/blob/cm_ref_backup/source/extensions/common/aws/utility.cc#L325

Reply:
For the most part I think the approach you want to take here is not too far from what you've done. But I suspect there's a better path that doesn't involve using global or static variables.

If your config functions in aws_iam/config.cc can be made part of a class, and that can be constructed at configuration time, then I think it'll be easier for you to plumb the CM in that way, rather than via grpc.

I don't totally understand why the CM is needed for you to bypass curl, but if you did need the CM I think you could get it when you construct your aws utility object.

Right now you might not *have* an AWS utility object (I haven't checked) but I suspect that would make things easier. Or maybe there's another context you can latch onto that you control.

Reply:
As of today for aws utility we are not creating any objects. We are just utilizing static functions inside `aws::Utility` (ref: [utility.h](https://github.com/envoyproxy/envoy/blob/main/source/extensions/common/aws/utility.h)).
I am not able to readily find a `context` that I can utilize inside aws utility. This approach would have been much simpler.

Do you happen to know an example where cm is obtained from a context?

---

I don't have much knowledge on the core components of envoy code, but would making use of `Registry::FactoryRegistry` acceptable? I don't have a solution in mind, I just want to make sure if I can proceed in this direction.

> why the CM is needed

As of now I am trying to model aws metadata fetcher similar to [jwks_fetcher.cc](https://github.com/envoyproxy/envoy/blob/main/source/extensions/filters/http/common/jwks_fetcher.cc#L40). The CM will be used to get the threadLocalCluster to make httpAsync calls.